src/middleware.ts

```diff
@@ -258,36 +263,63 @@ export const persist = <S extends State>(
     void setItem()
   }
 
+  const makeThenable = (value: any) => {
```

Comment:
Does it work for error handling.
It has to be something like this? not 100% sure.
```ts
type Thenable<Result> = {
  then: (onFulfilled: ..., onRejected: ...) => ...
}
const makeFunctionThatReturnsThenable = <Input, Result>(
  fn: ((input: Input) => Result | Promise<Result>)
): (input: Input) => Thenable<Result> => {
 ...
}
```
Also, we might want to make types correctly to avoid misunderstanding.

Reply:
👍 to proper typing

I guess this would be the optimal solution if it weren't for the bundle size: https://github.com/fluffynuts/synchronous-promise. Might still serve as reference point, though.

Reply:
I'd like to avoid dependencies not only because of the size, but also other various bundling issues.

Reply:
it works for async error handling.
but for non async it won't.

Reply:
I will work on types after these changes become final.

Reply:
Ok, I am going to check how promise internally work. maybe I can improve `toThenable()` to exactly handle `then` and `catch` the same way as promise do.

Reply:
maybe `mongoose` source code has something to help.

Reply:
`then: (onResolve, onReject)`
will not work on a custom object

Reply:
I am working on a new solution that will handle catch on both async and sync. just a little debugging remained